File: sdks/nim/src/mongo_do.nim
## MongoDB SDK for the .do platform
## Natural Language First, AI-Native
##
## Example usage:
## ```nim
## import mongo_do
##
## # Natural language queries
## let users = mongo"users who haven't logged in this month"
## let vips = mongo"customers with orders over $1000"
##
## # MongoDB compatible API
## let client = newMongoClient("https://db.example.com")
## let db = client.database("myapp")
## let coll = db.collection("users")
##
## let alice = coll.findOne(%*{"email": "alice@example.com"})
## ```

import std/[asyncdispatch, json, tables, options, strutils, sequtils, locks, times, uri, httpclient, random]

import ./mongo_do/errors
import ./mongo_do/config
import ./mongo_do/cursor
import ./mongo_do/results
import ./mongo_do/query
import ./mongo_do/collection
import ./mongo_do/database
import ./mongo_do/client

export errors, config, cursor, results, query, collection, database, client
export asyncdispatch, json, options, tables, times

const Version* = "0.1.0"

# Global configuration
var globalConfig {.threadvar.}: MongoConfig

proc getGlobalConfig*(): MongoConfig =
  if globalConfig.isNil:
    globalConfig = newMongoConfig()
  result = globalConfig

proc configure*(config: MongoConfig) =
  ## Set global configuration
  globalConfig = config

# ---------------------------------------------------------
# Natural Language Query Macro
# ---------------------------------------------------------

template mongo*(queryString: static[string]): MongoQuery =
  ## Natural language query prefix
  ##
  ## Example:
  ## ```nim
  ## let inactive = mongo"users who haven't logged in this month"
  ## let vips = mongo"customers with orders over $1000"
  ## ```
  newMongoQuery(queryString, getGlobalConfig())

template mongo*(queryString: string): MongoQuery =
  ## Natural language query (runtime string)
  newMongoQuery(queryString, getGlobalConfig())

# ---------------------------------------------------------
# Transaction Support
# ---------------------------------------------------------

type
  Transaction* = ref object
    config*: MongoConfig
    operations: seq[JsonNode]
    committed: bool

proc newTransaction*(config: MongoConfig = nil): Transaction =
  Transaction(
    config: if config.isNil: getGlobalConfig() else: config,
    operations: @[],
    committed: false
  )

proc query*(tx: Transaction, queryString: string): MongoQuery =
  ## Execute a query within a transaction
  newMongoQuery(queryString, tx.config, tx)

proc addOperation*(tx: Transaction, op: JsonNode) =
  if not tx.committed:
    tx.operations.add(op)

proc commit*(tx: Transaction) =
  ## Commit the transaction
  if tx.committed:
    return
  # In real implementation, send operations atomically
  tx.committed = true

proc rollback*(tx: Transaction) =
  ## Rollback the transaction
  tx.operations = @[]
  tx.committed = true

template transaction*(body: untyped): untyped =
  ## Execute operations in a transaction
  ##
  ## Example:
  ## ```nim
  ## transaction:
  ##   query("alice account").debit(100)
  ##   query("bob account").credit(100)
  ## ```
  block:
    let tx {.inject.} = newTransaction()
    try:
      body
      tx.commit()
    except CatchableError as e:
      tx.rollback()
      raise e

File: sdks/elixir/lib/mongo_do.ex
defmodule MongoDo do
  @moduledoc """
  MongoDB on the Edge. Natural Language First. AI-Native.

  ## Quick Start

      import MongoDo

      # Natural language queries
      users = mongo("users who haven't logged in this month")
      vips = mongo("customers with orders over $1000")

      # Chain with pipes
      result =
        mongo("users in Austin")
        |> MongoDo.rmap(fn user -> mongo("recent orders for \#{user}") end)
        |> MongoDo.rmap(fn orders -> mongo("shipping status for \#{orders}") end)

  ## Configuration

      config :mongo_do,
        url: "https://your-worker.workers.dev",
        api_key: "your-api-key"
  """

  alias MongoDo.{Client, Connection, Query, Promise}

  @type query_result :: term() | {:error, term()}
  @type query_opts :: keyword()

  @doc """
  Execute a natural language MongoDB query.

  ## Examples

      mongo("users who haven't logged in this month")
      mongo("customers with orders over $1000")
      mongo("most popular products this week")
  """
  @spec mongo(String.t()) :: query_result()
  defmacro mongo(query) do
    quote do
      MongoDo.execute(unquote(query))
    end
  end

  @doc """
  Execute a natural language MongoDB query with options.

  ## Options

    * `:timeout` - Query timeout in milliseconds (default: 30_000)
    * `:pool` - Connection pool name (default: :default)

  ## Examples

      mongo("users in Austin", timeout: 60_000)
  """
  @spec mongo(String.t(), query_opts()) :: query_result()
  defmacro mongo(query, opts) do
    quote do
      MongoDo.execute(unquote(query), unquote(opts))
    end
  end

  @doc """
  Execute a natural language query, raising on error.

  ## Examples

      result = mongo!("users who haven't logged in this month")
  """
  @spec mongo!(String.t()) :: term()
  defmacro mongo!(query) do
    quote do
      case MongoDo.execute(unquote(query)) do
        {:error, error} -> raise MongoDo.QueryError, error
        result -> result
      end
    end
  end

  @doc """
  Execute a query (internal implementation).
  """
  @spec execute(String.t(), query_opts()) :: query_result()
  def execute(query, opts \\ []) do
    pool = Keyword.get(opts, :pool, :default)
    timeout = Keyword.get(opts, :timeout, 30_000)

    Connection.execute(pool, {:natural_query, query}, timeout)
  end

  @doc """
  Map over a promise result with a function.

  Used for promise pipelining - chains operations with minimal round trips.

  ## Examples

      mongo("active users")
      |> MongoDo.rmap(fn u -> mongo("pending orders for \#{u.id}") end)
      |> MongoDo.rmap(fn o -> o.total end)
      |> MongoDo.await()
  """
  @spec rmap(Promise.t() | term(), (term() -> term())) :: Promise.t()
  def rmap(promise_or_result, fun) do
    Promise.map(promise_or_result, fun)
  end

  @doc """
  Await a promise, executing the pipeline and returning the result.

  ## Examples

      result =
        mongo("active users")
        |> MongoDo.rmap(fn u -> mongo("orders for \#{u}") end)
        |> MongoDo.await()
  """
  @spec await(Promise.t()) :: term()
  def await(promise) do
    Promise.await(promise)
  end

  @doc """
  Execute a transaction with a function.

  All operations within the function are executed atomically.

  ## Examples

      MongoDo.transaction(fn tx ->
        tx |> query("alice account") |> debit(100)
        tx |> query("bob account") |> credit(100)
      end)
  """
  @spec transaction((MongoDo.Transaction.t() -> term())) :: {:ok, term()} | {:error, term()}
  def transaction(fun) when is_function(fun, 1) do
    MongoDo.Transaction.execute(fun)
  end

  @doc """
  Connect to a MongoDB instance.

  ## Examples

      {:ok, client} = MongoDo.connect("https://your-worker.workers.dev")
  """
  @spec connect(String.t(), keyword()) :: {:ok, Client.t()} | {:error, term()}
  def connect(uri, opts \\ []) do
    Client.connect(uri, opts)
  end

  @doc """
  Get a database from a client.

  ## Examples

      db = MongoDo.database(client, "myapp")
  """
  @spec database(Client.t(), String.t()) :: MongoDo.Database.t()
  def database(client, name) do
    Client.database(client, name)
  end
end
